### PR TITLE
feat: add keys into kafka header while producing

### PR DESCRIPTION
**Pull request by vigith**
partially solves: #2142 

we will be adding `__key_len` in the header to indicate how many keys are going to be in the header. `__key_%d` will have the value at an index.

### CODE REVIEW

File: pkg/sinks/kafka/kafka.go [hunk 1]
```diff
@@ -150,10 +150,32 @@ func (tk *ToKafka) Write(_ context.Context, messages []isb.Message) ([]isb.Offse
...
+			Key:   []byte("__key_len"),
+			Value: []byte(fmt.Sprintf("%d", len(msg.Keys))),
+		}
+		headers = append(headers, keyLen)
+
+		// write keys into header if length > 0
+		if len(msg.Keys) > 0 {
+			for idx, key := range msg.Keys {
+				headers = append(headers, sarama.RecordHeader{
+					Key:   []byte(fmt.Sprintf("__key_%d", idx)),
```
Comment (whynowy) [inline pkg/sinks/kafka/kafka.go:169]: Document this in the Kafka sink? https://numaflow.numaproj.io/user-guide/sinks/kafka/

**Review by whynowy — APPROVED** [review]

File: pkg/sinks/kafka/kafka.go (hunk 1 above)
Comment (vegesana) [reply pkg/sinks/kafka/kafka.go:169]: Hi,

filling Headers in the producer message is good. This is needed for retaining the headers. But my requirement is the **"key"** part of producer message. My understanding is that this key will decide on which partition of kafka the message will be put in. 

type ProducerMessage struct {
	Topic string // The Kafka topic for this message.
	// The partitioning key for this message. Pre-existing Encoders include
	// StringEncoder and ByteEncoder.
	Key Encoder                 **<< THis is the one i am looking for**
	// The actual message to store in Kafka. Pre-existing Encoders include
	// StringEncoder and ByteEncoder.
	Value Encoder

	// The headers are key-value pairs that are transparently passed
	// by Kafka between producers and consumers.
	Headers []RecordHeader

**Review by vegesana — CHANGES_REQUESTED** [review]
Need a way to send the key value that is added in UDF to be retained till the end where kafka sink produces into the topic.

File: pkg/sinks/kafka/kafka.go (hunk 1 above)
Comment (vigith) [reply pkg/sinks/kafka/kafka.go:169]: i will add via a separate PR.

**Pull request merged**